pkg/nsx/services/inventory/build.go

```diff
@@ -201,3 +202,137 @@ func (s *InventoryService) BuildNamespace(namespace *corev1.Namespace) (retry bo
+		Tags:               GetTagsFromLabels(service.Labels),
+		ContainerClusterId: s.NSXConfig.Cluster,
+		ContainerProjectId: string(namespace.UID),
+		ExternalId:         string(service.UID),
+		NetworkErrors:      nil,
+		NetworkStatus:      netStatus,
+		OriginProperties:   originProperties,
+		Status:             status,
+	}
+
+	log.V(1).Info("Build service", "current application", containerApplication, "pre application", preContainerApplication)
+	operation, _ := s.compareAndMergeUpdate(preContainerApplication, containerApplication)
+	if operation != operationNone {
+		s.pendingAdd[containerApplication.ExternalId] = &containerApplication
+	} else {
+		log.Info("Skip, service not updated", "Service", service.Name, "Namespace", service.Namespace)
+	}
+	return
+}
+
+func (s *InventoryService) updateContainerApplicationInstanceServiceID(podUID string, service *corev1.Service) (retry bool) {
+	applicationInstance := s.ApplicationInstanceStore.GetByKey(podUID)
+	if applicationInstance == nil {
+		log.Info("Endpoint for pod %s is before pod creation, retry service %s to establish correlation", podUID, service.Name)
+		return true
+	} else {
+		allInstances := s.ApplicationInstanceStore.List()
+		// Ensure that adding multiple services can work in a batch.
+		if s.pendingAdd[podUID] != nil {
+			applicationInstance = s.pendingAdd[podUID]
+		}
+		ctx := context.TODO()
+		pod, err := GetPodByUID(ctx, s.Client, types.UID(podUID), service.Namespace)
+		if err != nil {
+			log.Error(err, "Failed to get Pod by UID", "PodUID", podUID, "Namespace", service.Namespace)
+			return true
+		}
+		serviceUIDs, err := GetServicesUIDByPodUID(ctx, s.Client, pod.UID, pod.Namespace)
+		if err != nil {
+			log.Error(err, "Failed to get services UIDs by pod UID", "Pod UID", pod.UID, "Namespace", pod.Namespace)
+			return true
+		}
+		updatedContainerApplicationInstance := applicationInstance.(*containerinventory.ContainerApplicationInstance)
+		updatedContainerApplicationInstance.ContainerApplicationIds = serviceUIDs
+		diff := make(map[string]interface{})
+		diff["external_id"] = updatedContainerApplicationInstance.ExternalId
+		diff["resource_type"] = string(ContainerApplicationInstance)
+		diff["container_application_ids"] = serviceUIDs
+		s.requestBuffer = append(s.requestBuffer, containerinventory.ContainerInventoryObject{ContainerObject: diff, ObjectUpdateType: operationUpdate})
+		s.pendingAdd[updatedContainerApplicationInstance.ExternalId] = updatedContainerApplicationInstance
+
+		// Remove stale service IDs from other ContainerApplicationInstances
+		for _, instObj := range allInstances {
+			inst := instObj.(*containerinventory.ContainerApplicationInstance)
+			if inst.ExternalId == podUID {
+				continue
+			}
+			// Remove any ContainerApplicationIds that are not in the new serviceUIDs
+			var newIds []string
+			for _, id := range inst.ContainerApplicationIds {
+				if !util.Contains(serviceUIDs, id) {
+					newIds = append(newIds, id)
+				}
+			}
```

Comment:
Could you double check if this works?
Suppose the podUID belongs to pod A, and it has a service called service A.
Here we list all the pods, saying we find a Pod B previously related to this service A.
Are we just trying to remove the service A from pod B, but it is possible that one service can connect to multiple pods?

Reply:
fixed